035-cluster-wide-volume-usage-quota-managemnet.md

```diff
@@ -0,0 +1,280 @@
+will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied. The broker will go offline
+if all volumes are unavailable.
+
+Configuring the soft and hard limit thresholds is currently done in terms of bytes consumed, which forces users
+deploying the plugin to calculate a threshold for throttling and would require separate thresholds per volume to avoid
+out of space issues.
+
+## Proposal
+### Leverage Kafka to distribute volume usage metrics throughout the cluster.
+By publishing per volume usage metrics to a compacted topic keyed by the broker ID, each instance of the plugin will be able to start and quickly determine the state of the cluster and consistently apply throttling regardless of where the
+client is connected.
+
+[KIP-257](https://cwiki.apache.org/confluence/display/KAFKA/KIP-257+-+Configurable+Quota+Management) describes configurable quota management, where quotas are defined in terms of how much of a delay to apply to a given produce request. This leads to defining the
+possible states of the plugin as:
+`OPEN`, `THROTTLE` and `PAUSE` to accurately reflect the effects of transitioning between states.
+
+This implies there are three roles.
+1. **Data source** - Publish volume usage metrics about the volumes underlying the brokers log dirs.
+2. **Data Sink** - Consume volume usage metrics and make those available for making Quota Policy decisions.
+3. **Quota Policy** - Convert volume usage metrics for the entire cluster into throttling decisions.
+
+This proposal envisages them all being implemented by the kafka-static-quota-plugin however by clearly distinguishing the roles we leave open future options to move each of the roles out of process.
+
+#### Limit Types
+1. Consumed space limit: Triggers if the consumed space of a volume breaches the value. Candidate for deprecation.
+2. Minimum Free Bytes: Triggers if the amount of free space on a volume drops below the configured level. Particularly
+   useful for hard limits as an absolute minimum of free space.
+3. Minimum Free Percentage. Triggers if the amount of free space on the volume drops below the configured proportion of
+   the volumes total capacity.
+
+#### The data source should:
+- Collect local volume usage
+- Publish usage to an internal compacted topic keyed by broker ID.
+- Collect and publish volume usage metrics (using this [message schema](#Message-schema)) on start up and periodically after that.
+- Connect as an internal service using TLS and client certificates to connect to the Replication listener.
+
+#### The data sink should:
+- Connect as an internal service using TLS and client certificates to connect to the Replication listener.
+- On startup and periodically thereafter read the volume usage stats from the topic
+- Use the Kafka Admin client to discover the list of currently active brokers
```

Comment:
We'll also need to use `describeLogDirs()` to find the logdir where each replica lives.